gio/src/subclass/async_initable.rs

```diff
@@ -0,0 +1,522 @@
+        #[inline]
+        pub(super) fn is_valid(&self, source_object: &AsyncInitable) -> bool {
+            self.imp().inner.get().unwrap().source_object == *source_object
+        }
+        #[inline]
+        pub(super) fn propagate_error(&self) -> Option<glib::Error> {
+            self.imp().error.take()
+        }
+    }
+
+    #[derive(Debug)]
+    struct Inner {
+        source_object: AsyncInitable,
+        user_data: glib::Pointer,
+        source_tag: glib::Pointer,
+    }
+
+    #[derive(Default)]
+    pub(super) struct AsyncInitableResult {
+        inner: once_cell::sync::OnceCell<Inner>,
+        error: std::cell::RefCell<Option<glib::Error>>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for AsyncInitableResult {
+        const NAME: &'static str = "GRustAsyncInitableResult";
+        type Type = super::AsyncInitableResult;
+        type Interfaces = (crate::AsyncResult,);
+    }
+
+    impl ObjectImpl for AsyncInitableResult {}
+
+    impl AsyncResultImpl for AsyncInitableResult {
+        #[inline]
+        fn user_data(&self, _res: &Self::Type) -> glib::Pointer {
+            self.inner.get().unwrap().user_data
+        }
+        #[inline]
+        fn source_object(&self, _res: &Self::Type) -> Option<glib::Object> {
+            Some(self.inner.get().unwrap().source_object.clone().upcast())
+        }
+        #[inline]
+        fn is_tagged(&self, _res: &Self::Type, source_tag: glib::Pointer) -> bool {
+            self.inner.get().unwrap().source_tag == source_tag
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::prelude::*;
+    use crate::traits::AsyncInitableExt;
+
+    pub mod imp {
+        use super::*;
+        use crate::AsyncInitable;
+        use std::cell::Cell;
+
+        pub struct AsyncInitableTestType(pub Cell<u64>);
+
+        #[glib::object_subclass]
+        impl ObjectSubclass for AsyncInitableTestType {
+            const NAME: &'static str = "AsyncInitableTestType";
+            type Type = super::AsyncInitableTestType;
+            type Interfaces = (AsyncInitable,);
+
+            fn new() -> Self {
+                Self(Cell::new(0))
+            }
+        }
+
+        impl AsyncInitableImpl for AsyncInitableTestType {
+            fn init_future(
+                &self,
+                initable: &Self::Type,
+                _io_priority: glib::Priority,
+            ) -> Pin<Box<dyn Future<Output = Result<(), Error>> + 'static>> {
+                let initable = initable.clone();
+                Box::pin(async move {
+                    glib::timeout_future_seconds(0).await;
+                    initable.imp().0.set(0x123456789abcdef);
+                    Ok(())
+                })
+            }
+        }
+
+        impl ObjectImpl for AsyncInitableTestType {}
+    }
+
+    pub mod ffi {
+        use super::*;
+        pub type AsyncInitableTestType = <imp::AsyncInitableTestType as ObjectSubclass>::Instance;
+
+        #[no_mangle]
```

Comment:
Why `no_mangle`? You're not linking to this from C

Reply:
Not sure, this was copied verbatim from the `Initable` test

Reply:
Let's remove it in both then